src/playground-code-editor.ts

```diff
@@ -181,6 +182,18 @@ export class PlaygroundCodeEditor extends LitElement {
     this.requestUpdate('value', oldValue);
   }
 
+  /**
+   * Alternative to `value`, and takes precedence over `value`. Use `docCache`
+   * to pass a CodeMirror document instance directly. If `docCache` is set,
+   * `value` will no longer update the contents of the editor. The document
+   * instance should be used if history needs to be preserved.
+   *
+   * Whenever a new document instance is created, a `newdocinstance` event is
+   * fired containing a `codeMirrorDoc` property.
+   */
+  @property({attribute: false})
```

Comment:
What do you think about this slight variation, which shifts the responsibility for the actual codemirror document caching into the hidden details of this module?

- Add a property `@property({attribute: false} documentKey?: object`
- Add a property `private readonly _documentCache = new WeakMap<object, Doc>()`.
- When a new `documentKey` is seen, create a new CodeMirror doc, and add it to the weakmap, and swap to it.
- When a known `documentKey` is set, get it from the cache and swap to it.

Then in `file-editor`, we can just have a binding like:

```ts
<playground-code-editor .documentKey=${this._currentFile}>
```

So, the identity of the `_currentFile` object itself serves as the key, and `code-editor` handles all of the details. The `WeakMap` is used so that the `code-editor` doesn't become a source of memory leaks when deleting files, switching projects, etc.

I think some advantages of this could be:

- Slightly simpler API. Only the one new public property is needed.

- Preserves the characteristic of not exposing any CodeMirror implementation details to public API. I've been thinking this is a nice thing to preserve, in particular because it would make it really easy to have drop-in replacements for `<playground-code-editor>` (such as a Monaco implementation), without `<playground-file-editor>` needing to know the difference.

It could also be nice as part of this to change the behavior of `value` when there is no `documentKey`:

- If `documentKey` is `undefined`, then whenever `value` is set *from the outside*, always create a new CodeMirror doc and swap to it. I think this is a better default than what we have now, because it's worse to have incorrect undo history, than no undo history at all.

Reply:
Chose not to include the breaking change when no `documentKey` is present. I think it's reasonable to have a single history stack by default. And it's only when adding abstractions like a file system where the `documentKey` becomes useful.